cpp/include/raft/spatial/knn/detail/ball_cover/registers.cuh

```diff
@@ -0,0 +1,514 @@
+    if (cur_R_dist > weight * (min_R_dist + R_radius[cur_R_ind])) continue;
+    if (cur_R_dist > 3 * min_R_dist) return;
+
+    // The whole warp should iterate through the elements in the current R
+    value_idx R_start_offset = R_indptr[cur_R_ind];
+    value_idx R_stop_offset = R_indptr[cur_R_ind + 1];
+
+    value_idx R_size = R_stop_offset - R_start_offset;
+
+    int limit = faiss::gpu::utils::roundDown(R_size, faiss::gpu::kWarpSize);
+    int i = threadIdx.x;
+    for (; i < limit; i += tpb) {
+      // Index and distance of current candidate's nearest landmark
+      value_idx cur_candidate_ind = R_1nn_cols[R_start_offset + i];
+      value_t cur_candidate_dist = R_1nn_dists[R_start_offset + i];
+
+      // Take 2 landmarks l_1 and l_2 where l_1 is the furthest point in the heap
+      // and l_2 is the current landmark R. s is the current data point and
+      // t is the new candidate data point. We know that:
+      // d(s, t) cannot possibly be any smaller than | d(s, l_1) - d(l_1, l_2) | * | d(l_1, l_2) - d(l_2, t) | - d(s, l_1) * d(l_2, t)
+
+      // Therefore, if d(s, t) >= d(s, l_1) from the computation above, we know that the distance to the candidate point
+      // cannot possibly be in the nearest neighbors. However, if d(s, t) < d(s, l_1) then we should compute the
+      // distance because it's possible it could be smaller.
+      //
+      value_t z = heap.warpKTopRDist == 0.00
+                    ? 0.0
+                    : (abs(heap.warpKTop - heap.warpKTopRDist) *
+                         abs(heap.warpKTopRDist - cur_candidate_dist) -
+                       heap.warpKTop * cur_candidate_dist) /
+                        heap.warpKTopRDist;
+
+      z = isnan(z) ? 0.0 : z;
+      value_t dist = std::numeric_limits<value_t>::max();
+      if (i < k || z <= heap.warpKTop) {
+        const value_t *y_ptr = X_index + (n_cols * cur_candidate_ind);
+        value_t local_y_ptr[col_q];
+        for (int j = 0; j < n_cols; j++) local_y_ptr[j] = y_ptr[j];
+        dist = dfunc(local_x_ptr, local_y_ptr, n_cols);
+        ++n_dists_computed;
+      }
+
+      heap.add(dist, cur_candidate_dist, cur_candidate_ind);
+    }
+
+    if (i < R_size) {
+      value_idx cur_candidate_ind = R_1nn_cols[R_start_offset + i];
+      value_t cur_candidate_dist = R_1nn_dists[R_start_offset + i];
+      value_t z = heap.warpKTopRDist == 0.0
+                    ? 0.0
+                    : (abs(heap.warpKTop - heap.warpKTopRDist) *
+                         abs(heap.warpKTopRDist - cur_candidate_dist) -
+                       heap.warpKTop * cur_candidate_dist) /
+                        heap.warpKTopRDist;
+
+      z = isnan(z) ? 0.0 : z;
+      value_t dist = std::numeric_limits<value_t>::max();
+      if (i < k || z <= heap.warpKTop) {
+        const value_t *y_ptr = X_index + (n_cols * cur_candidate_ind);
+        value_t local_y_ptr[col_q];
+        for (int j = 0; j < n_cols; j++) local_y_ptr[j] = y_ptr[j];
```

Comment:
I won't keep pointing it out, but in general I think it would be good to put braces around single-line `for` and  `if` clauses to reduce the likelihood of future developers accidentally pushing code that should be in the clause out of it. Also, don't forget to use pre-increment unless we're making use of the output value.

Reply:
> Also, don't forget to use pre-increment unless we're making use of the output value.

Is this necessary in a for loop?

Reply:
Ah, I guess it can save the need to explicitly store the value for one iteration in memory.

Reply:
"Necessary" is a strong word :stuck_out_tongue:. Godbolt reports that gcc compiles them to the same assembly, but in general, I think it's preferable to be explicit about when you do and don't want the temporary that would implicitly get created with post-increment without optimization. If I'm using post-increment, it's a signal to future developers that yes, I really did mean to use that return value and if I didn't make use of it, it's a mistake.